gensim/corpora/text_processing_pool.py

```diff
@@ -0,0 +1,160 @@
+    def process(self, text):
+        """Process a document text; subclasses should override."""
+        raise NotImplementedError
+
+    def run(self):
+        inqueue, outqueue, initializer, initargs, maxtasks = self._args
+
+        assert maxtasks is None or (type(maxtasks) == int and maxtasks > 0)
+        put = outqueue.put
+        get = inqueue.get
+        if hasattr(inqueue, '_writer'):
+            inqueue._writer.close()
+            outqueue._reader.close()
+
+        if initializer is not None:
+            initializer(*initargs)
+
+        completed = 0
+        while maxtasks is None or (maxtasks and completed < maxtasks):
+            try:
+                task = get()
+            except (EOFError, OSError):
+                util.debug('worker got EOFError or OSError -- exiting')
+                break
+
+            if task is None:
+                util.debug('worker got sentinel -- exiting')
+                break
+
+            job, i, func, args, kwds = task
+            try:
+                if func is mapstar:
+                    args = [(self.process, arg_list) for _, arg_list in args]
+                else:
+                    func = self.process
+                result = (True, func(*args, **kwds))
+            except Exception as e:
+                result = (False, e)
+            try:
+                put((job, i, result))
+            except Exception as e:
+                wrapped = MaybeEncodingError(e, result[1])
+                util.debug("Possible encoding error while sending result: %s" % (
+                    wrapped))
+                put((job, i, (False, wrapped)))
+
+            task = job = result = func = args = kwds = None
+            completed += 1
+        util.debug('worker exiting after %d tasks' % completed)
+
+
+class _PatchedPool(mp.pool.Pool):
+
+    def Process(self, *args, **kwds):
+        kwds['state_kwargs'] = self._state_kwargs
+        return self._processor_class(*args, **kwds)
+
+    def __init__(self, processes=None, initializer=None, initargs=(), maxtasksperchild=None,
+                 state_kwargs=None, processor_class=TextProcessor):
+        self._state_kwargs = state_kwargs
+        self._processor_class = processor_class
+        super(_PatchedPool, self).__init__(processes, initializer, initargs, maxtasksperchild)
+
+    def _repopulate_pool(self):
+        """Bring the number of pool processes up to the specified number,
+        for use after reaping workers which have exited.
+        """
+        for i in range(self._processes - len(self._pool)):
+            w = self.Process(args=(self._inqueue, self._outqueue,
+                                   self._initializer,
+                                   self._initargs, self._maxtasksperchild))
+            self._pool.append(w)
+            w.name = w.name.replace('Process', 'PoolWorker')
+            w.daemon = True
+            w.start()
+            util.debug('added worker')
+
+
+class TextProcessingPool(object):
```

Comment:
Added documentation with rationale. See comment on `_PatchedPool` above for additional context.